keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -0,0 +1,248 @@
+    * kubernetes tooling
+      * minimal change
+    * compatibility implications
+      * breaking go changes in each release impact consumers that have not pinned to particular tags/shas (as they do today)
+      * conflicting version requirements (direct or transitive) can result in impossible-to-build or impossible-to-update dependencies (as they do today)
+
+2. tag major versions on every `kubernetes/kubernetes` release (similar to what client-go does), and use semantic import versioning
+
+    * consumers
+      * published modules import as a consistent version (e.g. `v15.0.0`)
+      * `go get` behavior (e.g. `go get client-go@v15.0.0`):
+        * uses referenced version of `k8s.io/...` dependencies
+        * uses referenced commit of transitive non-module-based dependencies (unless overridden by top-level module or conflicting peers)
+      * GOPATH consumers import `k8s.io/apimachinery/...` (as they do today)? (needs research to verify the internal versioned imports can build in GOPATH mode)
+      * module-based consumers import `k8s.io/apimachinery/v15/...`
+      * module-based consumers can have multiple copies of kubernetes libraries (though non-module transitive dependencies could still conflict)
+    * kubernetes tooling
+      * requires rewriting all `k8s.io/{component}/vX/...` imports at the start of each release
+      * requires updating code generation scripts to generate versioned imports for `k8s.io/{api,apimachinery,client-go}`, etc
+      * might require changes to build scripts to include version in built package path (needs research to verify)
+    * compatibility implications
+      * allows breaking go changes in each kubernetes "minor" release
+      * no breaking go changes are allowed in a kubernetes patch releases (need tooling to enforce this)
+
+3. tag minor versions on every `kubernetes/kubernetes` release
+
+    * consumers
+      * published modules import as a consistent version (e.g. `v1.15.0`)
+      * `go get` behavior (e.g. `go get client-go@v15.0.0`):
+        * pulls latest minor version of transitive `k8s.io/...` dependencies (similar to pulling latest master)
+        * pulls latest commit of transitive non-module-based dependencies (likely to break)
+      * GOPATH consumers import `k8s.io/apimachinery/...` (as they do today)
+      * module-based consumers import `k8s.io/apimachinery/...` (as they do today)
+      * consumers are limited to a single copy of kubernetes libraries among all dependencies
+    * kubernetes tooling
+      * minimal change
+    * compatibility requirements
+      * requires strict go compatibility with all tagged releases for the current major version (need tooling to enforce this)
+      * no breaking changes: moved/removed/changed fields/types/methods/interfaces
+      * needs research:
+        * is there a tool to check go API compatibility between two commits?
+        * how compatible has each component been historically?
```

Comment:
the answer won't make us happy.

Reply:
my guess is that every staging component has broken go compatibility in some form on every release

Reply:
k8s/k8s doesn't follow semver on the go api. The staging projects that use the form `kubernetes-*` break semver on the go api as well in minor releases. Moving to semver would be a change.

Reply:
To ensure this is scoped small enough to be completed successfully in the 1.15 timeframe, I've proposed keeping the existing tagging/versioning scheme for the first iteration. This behaves no worse under go modules, and doesn't prevent us from moving to something like semantic import versioning in the future.